sshd-core/src/test/java/org/apache/sshd/client/ClientTest.java

```diff
@@ -1530,6 +1532,29 @@ public void testConnectUsingIPv6Address() throws IOException {
         }
     }
 
+    @Test // see SSHD-1295
+    public void testConnectTimeout() throws Exception {
+        client.start();
+        try {
+            ConnectFuture future = client.connect(getCurrentTestName(), TEST_LOCALHOST, port);
+            try {
+                future.verify(1);
+                fail("Timeout expected");
+            } catch (InterruptedIOException | SshException e) {
+                ClientSession session = null;
+                try {
+                    session = future.verify(CONNECT_TIMEOUT).getSession();
+                } catch (SshException e2) {
+                    assertTrue("Expected a timeout, got " + e2, e2.getMessage().contains("timeout"));
+                }
+
+                assertNull("Session should not set since client timeout", session);
```

Comment:
What is missing in this test now is a verification that the session, if one got created, was indeed closed.

Maybe install a SessionListener on the client before starting it. That sessionListener gets invoked whenever a ClientSession is created and can store that session. Then check here that all sessions that got created are closed or closing.

Reply:
Add before line 1537
```
        List<Session> sessions = new CopyOnWriteArrayList<>();
        client.addSessionListener(new SessionListener() {

            @Override
            public void sessionCreated(Session session) {
                // Delay a little bit to ensure that verify(1) does time out below
                try {
                    Thread.sleep(100);
                } catch (InterruptedException e) {
                    Thread.currentThread().interrupt();
                }
                sessions.add(session);
            }
        });
```
and then after line 1551
```
                for (Session created : sessions) {
                    assertTrue("Created session should be closed", created.isClosed() || created.isClosing());
                }
```

Reply:
Thank you. By the way, why does it work locally but fail in GitHub Action? And need to add the delay operation.

Reply:
It's a timing issue. When I fetched your change and ran the test the first time, it worked locally, but I had debug logging on. When I switched off debug logging, it failed for me locally, too.